### ISSUE TITLE
RCLONE MOUNT -  Move & DirMove on unsupported backends (mimic with copy/remove)

### ISSUE THREAD

**Issue by tcf909** [ncw/rclone#2539]
Hello,

For a mounted backend (rclone mount) would there be any way to allow a "virtual move" for backends that do not support "Move" and "DirMove". Specifically: If the backend supports server side copy, a move command will copy then remove, rather than "move"?

This would really help out Google Cloud Storage (and others) and would be in line with the efforts around VFS (make the filesystem act more like a filesystem and less like a cloud store bucket).

Thoughts?

**Comment by tcf909** [comment]
I noticed on the website, this is the stated method of execution, but it does not actually work on google cloud storage.

rclone v1.43
- os/arch: darwin/amd64
- go version: go1.11

```
root@www-0:/shared/sites/digitalspaces.app# mkdir test
root@www-0:/shared/sites/digitalspaces.app# mv test test2
mv: cannot move 'test' to 'test2': Input/output error
```
```
2018/09/05 22:54:27 DEBUG : /: Attr:
2018/09/05 22:54:27 DEBUG : /: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2018/09/05 22:54:30 DEBUG : /: Lookup: name="test"
2018/09/05 22:54:30 DEBUG : : Re-reading directory (1m12.131528408s old)
2018/09/05 22:54:31 DEBUG : /: >Lookup: node=<nil>, err=no such file or directory
2018/09/05 22:54:31 DEBUG : /: Mkdir: name="test"
2018/09/05 22:54:31 DEBUG : /: >Mkdir: node=test/, err=<nil>
2018/09/05 22:54:31 DEBUG : test/: Attr:
2018/09/05 22:54:31 DEBUG : test/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2018/09/05 22:54:34 DEBUG : /: Lookup: name="test2"
2018/09/05 22:54:34 DEBUG : /: >Lookup: node=<nil>, err=no such file or directory
2018/09/05 22:54:34 DEBUG : test/: Attr:
2018/09/05 22:54:34 DEBUG : test/: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2018/09/05 22:54:34 DEBUG : /: Lookup: name="test2"
2018/09/05 22:54:34 DEBUG : /: >Lookup: node=<nil>, err=no such file or directory
2018/09/05 22:54:34 DEBUG : /: Lookup: name="test2"
2018/09/05 22:54:34 DEBUG : /: >Lookup: node=<nil>, err=no such file or directory
2018/09/05 22:54:34 DEBUG : /: Rename: oldName="test", newName="test2", newDir=/
2018/09/05 22:54:34 ERROR : test: Dir.Rename error: Fs "Storage bucket digital-spaces-www-local" can't rename directories (no DirMove)
2018/09/05 22:54:34 DEBUG : /: >Rename: err=Fs "Storage bucket digital-spaces-www-local" can't rename directories (no DirMove)
```

**Comment by tcf909** [comment]
Also tried without any cache options -- same result (failure).

**Comment by brechmos** [comment]
I think I am getting a similar error (with a dropbox mount).  I am moving one file on another that already exists.
```
$ mv nonce.tmp nonce.tmp2
2018/09/07 22:10:25 ERROR : nonce.tmp: File.Rename error: move failed: to/conflict/file/... 2018/09/07 22:10:25 ERROR : nonce.tmp: Dir.Rename error: move failed: to/conflict/file/... mv: cannot move 'nonce.tmp' to 'nonce.tmp2': Input/output error
```

My mount options are `uid=1000,gid=1000,allow-other,default-permissions,max-read-ahead=16M,_netdev` (using the rclonefs script from the wiki).

**Comment by ncw** [comment]
I think this issue is a duplicate of #1965 which is scheduled to be done soon.

@brechmos I think renaming should work on dropbox - do you get that problem every time?

**Comment by brechmos** [comment]
Sorry, yeah, I updated my comment above noting the #1965 connection but it was after I first posted :/

I just tried with dropbox and I **am** getting an error. 

This is the mounting:
```
$ ~/bin/rclone mount -v --debug-fuse  Dropbox: /tmp/db
2018/09/11 07:44:27 INFO  : Dropbox root '': poll-interval is not supported by this remote
2018/09/11 07:44:37 ERROR : Test/a: File.Rename error: move failed: to/conflict/file/..
2018/09/11 07:44:37 ERROR : Test/a: Dir.Rename error: move failed: to/conflict/file/..
```

The error above came when I did:
```
$ cd /tmp/db
$ touch a b  # Make two empty files.
$ mv a b
mv: cannot move 'a' to 'b': Input/output error
```

I am quite new at rclone, so there might be some helpful other flags but with the reasonably bare mount command I am seeing an error.

Oh... so, interesting. I just did the same thing with Mega and it worked:
```
$ ~/bin/rclone mount -v --debug-fuse  Mega: /tmp/db
2018/09/11 07:48:42 INFO  : mega root '': poll-interval is not supported by this remote
2018/09/11 07:48:56 INFO  : test/a: Copied (new)
2018/09/11 07:48:57 INFO  : test/b: Copied (new)
```

It also works with pCloud:
```
$ ~/bin/rclone mount -v --debug-fuse pCloud: /tmp/db
2018/09/11 07:50:05 INFO  : pcloud root '': poll-interval is not supported by this remote
2018/09/11 07:50:23 INFO  : test/a: Copied (new)
2018/09/11 07:50:23 INFO  : test/b: Copied (new)
```

It also worked with Google Drive. 

Tried Box and it did **not** work. 

Huh, so it seems to be **only** Box and Dropbox (of the ones I tested) that are giving me issues.

All of the above is based on:
```
rclone v1.43
- os/arch: linux/arm
- go version: go1.11
```

If you want me to test a PR, I am happy to do it.

**Comment by brechmos** [comment]
I tried a few more times in different incantations and it seems to be pretty consistent (for me).

**Comment by ncw** [comment]
@brechmos great testing.

Can you put your findings  in a new issue please - I think your problem is unrelated to this issue.

It seems to be that backends are inconsistent about whether they allow a file to be copied over or not.

**Comment by tcf909** [comment]
@ncw your right does look like a dupe. The difference is this is Google Cloud Storage. Not sure how your tracking things. (Per feature, per backend, etc.) Feel free to close.

Looking forward to this being done. Will make using mount much more like a real fs for those backends that don't support move.

**Comment by brechmos** [comment]
Per @ncw I created a ticket for my findings https://github.com/ncw/rclone/issues/2569

**Comment by ncw** [comment]
I think this issue is now fixed with `Move` falling back to Copy if not availabe.

`DirMove` does not fall back to `Move` or `Copy` yet though.  I'm uncertain as to whether that is a good idea or not as it might involve renaming 1000s of files for one directory rename.

**Comment by tcf909** [comment]
To make mount operate more like a real dir wouldn't we need to make sure
dirmove failed back? At least make it a flag like vfs so you can force a
more filesystem type experience?

**Comment by ncw** [comment]
@tcf909 wrote:
> To make mount operate more like a real dir wouldn't we need to make sure dirmove failed back? At least make it a flag like vfs so you can force a more filesystem type experience?

I guess it is a matter of tradeoffs...

To rename a directory rclone will

 1. Call `DirMove` if it is implemented (see [overview table](https://rclone.org/overview/#optional-features) for which backends do implement it)
  - works for 12/21 of backends
 2. Call `Move` on each file if available
  - works for another 0/21 backends (all the backends which implement `Move` implement `DirMove`)
 3. Call `Copy` then `Delete` on each file if available
  - works for another 6/21 backends
 4. Download then upload each file
  - works for the remaining 2/21 backends (b2 & yandex)

At the moment rclone mount will stop at step 1) above.  I think it could go down to step 3) possibly, but step 4) I think is too much, at least without a flag.

To put those into perspective let's say we have a directory with 1000 files and 100GB of data in. This is what rclone does for each of those steps

  1. Do one API call
  2. Do 1000 API calls
  3. Do 2000 API calls
  4. Do 3000 API calls, download 100GB and upload 100GB of data

What do you think rclone should do?

**Comment by tcf909** [comment]
I agree with your analysis. The final step is too much without a flag.

If you were able to cover the first three, and cover 22 of the 24 backends
that would be great.

**Comment by ncw** [comment]
Hmm, this is harder than it looks!  The obvious thing to do is re-use the internal syncing routines, however they don't take a directory path so they need to be re-worked first which is a much bigger job!

**Comment by dragetd** [comment]
There are applications which primarily save data but have certain config files with a move-operation on them. Those files are small and only a few - but the software fails without being able to mv them.

I tried running those on b2 and it fails. It would be awesome if rclone had a workaround for this, even knowing that it can be very expensive sometimes.

**Comment by dragetd** [comment]
I contacted b2 support a few days ago. They said they are thinking about implementing a move API during this year.

But I have given up on b2 anyways and just moved to wasabi.com - they offer similar prices and do support moving files around. :)
